perception/autoware_tensorrt_bevdet/src/bevdet.cpp

```diff
@@ -0,0 +1,717 @@
+    flag = 0;
+  }
+
+  // the smaller the idx, the newer th adj_bevfeat
+  for (int i = 0; i < adj_num_; i++) {
+    const void * adj_buffer = adj_frame_ptr_->getFrameBuffer(i);
+
+    size_t buf_size = trt_buffer_sizes_[buffer_map_["adj_feats"]] / adj_num_;
+
+    CHECK_CUDA(cudaMemcpy(
+      reinterpret_cast<char *>(trt_buffer_dev_[buffer_map_["adj_feats"]]) + i * buf_size,
+      adj_buffer, buf_size, cudaMemcpyDeviceToDevice));
+
+    Eigen::Quaternion<float> adj_ego2global_rot;
+    Eigen::Translation3f adj_ego2global_trans;
+    adj_frame_ptr_->getEgo2Global(i, adj_ego2global_rot, adj_ego2global_trans);
+
+    getCurr2AdjTransform(
+      ego2global_rot, adj_ego2global_rot, ego2global_trans, adj_ego2global_trans,
+      reinterpret_cast<float *>(trt_buffer_dev_[buffer_map_["transforms"]]) + i * transform_size_);
+  }
+  CHECK_CUDA(cudaMemcpy(
+    trt_buffer_dev_[buffer_map_["flag"]], &flag, trt_buffer_sizes_[buffer_map_["flag"]],
+    cudaMemcpyHostToDevice));
+}
+
+void BEVDet::getCurr2AdjTransform(
+  const Eigen::Quaternion<float> & curr_ego2global_rot,
+  const Eigen::Quaternion<float> & adj_ego2global_rot,
+  const Eigen::Translation3f & curr_ego2global_trans,
+  const Eigen::Translation3f & adj_ego2global_trans, float * transform_dev)
+{
+  Eigen::Matrix4f curr_e2g_transform;
+  Eigen::Matrix4f adj_e2g_transform;
+
+  for (int i = 0; i < 3; i++) {
+    for (int j = 0; j < 3; j++) {
+      curr_e2g_transform(i, j) = curr_ego2global_rot.matrix()(i, j);
+      adj_e2g_transform(i, j) = adj_ego2global_rot.matrix()(i, j);
+    }
+  }
+  for (int i = 0; i < 3; i++) {
+    curr_e2g_transform(i, 3) = curr_ego2global_trans.vector()(i);
+    adj_e2g_transform(i, 3) = adj_ego2global_trans.vector()(i);
+
+    curr_e2g_transform(3, i) = 0.f;
+    adj_e2g_transform(3, i) = 0.f;
+  }
+  curr_e2g_transform(3, 3) = 1.f;
+  adj_e2g_transform(3, 3) = 1.f;
+
+  Eigen::Matrix4f currEgo2adjEgo = adj_e2g_transform.inverse() * curr_e2g_transform;
+  Eigen::Matrix3f currEgo2adjEgo_2d;
+  for (int i = 0; i < 2; i++) {
+    for (int j = 0; j < 2; j++) {
+      currEgo2adjEgo_2d(i, j) = currEgo2adjEgo(i, j);
+    }
+  }
+  currEgo2adjEgo_2d(2, 0) = 0.f;
+  currEgo2adjEgo_2d(2, 1) = 0.f;
+  currEgo2adjEgo_2d(2, 2) = 1.f;
+  currEgo2adjEgo_2d(0, 2) = currEgo2adjEgo(0, 3);
+  currEgo2adjEgo_2d(1, 2) = currEgo2adjEgo(1, 3);
+
+  Eigen::Matrix3f gridbev2egobev;
+  gridbev2egobev(0, 0) = x_step_;
+  gridbev2egobev(1, 1) = y_step_;
+  gridbev2egobev(0, 2) = x_start_;
+  gridbev2egobev(1, 2) = y_start_;
+  gridbev2egobev(2, 2) = 1.f;
+
+  gridbev2egobev(0, 1) = 0.f;
+  gridbev2egobev(1, 0) = 0.f;
+  gridbev2egobev(2, 0) = 0.f;
+  gridbev2egobev(2, 1) = 0.f;
+
+  Eigen::Matrix3f currgrid2adjgrid = gridbev2egobev.inverse() * currEgo2adjEgo_2d * gridbev2egobev;
+
+  CHECK_CUDA(cudaMemcpy(
+    transform_dev, Eigen::Matrix3f(currgrid2adjgrid.transpose()).data(),
+    transform_size_ * sizeof(float), cudaMemcpyHostToDevice));
+}
+
+int BEVDet::doInfer(
```

Comment:
Namespaces: Wrap the function `doInfer` within the  `autoware_tensorrt_bevdet` namespace.